### ISSUE TITLE
Error on Install

### ISSUE THREAD

**Issue by arjunmehta** [cmake-js/fastcall#11]
Hi there, I'm trying to install another module (NOOOCL) actually, but I'm encountering an issue with fastcall installation.

I am using MacOS Sierra with XCode 8.1.
I've tried with various versions of Node (4.x-7.x) and NPM. I've even tried various various of node-gyp.

The fastcall build fails at `Building CXX object src/CMakeFiles/fastcall.dir/dyncallwrapper.cpp.o`.

```bash
Scanning dependencies of target fastcall
[ 51%] Building CXX object src/CMakeFiles/fastcall.dir/callbackimpl.cpp.o
[ 53%] Building CXX object src/CMakeFiles/fastcall.dir/dyncallbackwrapper.cpp.o
[ 56%] Building CXX object src/CMakeFiles/fastcall.dir/dyncallwrapper.cpp.o
/obscura/node_modules/fastcall/src/dyncallwrapper.cpp:584:5: error: no matching function for
      call to 'CallAsync'
    CallAsync<uint64_t>(
    ^~~~~~~~~~~~~~~~~~~
/obscura/node_modules/fastcall/src/dyncallwrapper.cpp:122:13: note: candidate function
      [with T = unsigned long long] not viable: no known conversion from '(lambda at
      /obscura/node_modules/fastcall/src/dyncallwrapper.cpp:587:9)' to 'typename
      CallAsyncWorker<unsigned long long>::TConvertFunc' (aka 'v8::Local<v8::Value> (*)(unsigned long long)') for 3rd
      argument
inline void CallAsync(
            ^
1 error generated.
```

Full output to that point:
```bash
info TOOL Using Unix Makefiles generator.
info TOOL Using c++11 compiler standard.
info CMD CONFIGURE
info RUN cmake "/obscura/node_modules/fastcall" --no-warn-unused-cli -G"Unix Makefiles" -DCMAKE_JS_VERSION="3.3.1" -DCMAKE_BUILD_TYPE="Release" -DCMAKE_LIBRARY_OUTPUT_DIRECTORY="/obscuras/node_modules/fastcall/build/Release" -DCMAKE_JS_INC="/Users/me/.cmake-js/node-x64/v7.2.1/include/node;/obscura/node_modules/nan" -DNODE_RUNTIME="node" -DNODE_RUNTIMEVERSION="7.2.1" -DNODE_ARCH="x64" -DCMAKE_CXX_FLAGS="-std=c++11 -D_DARWIN_USE_64_BIT_INODE=1 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DBUILDING_NODE_EXTENSION -w" -DCMAKE_SHARED_LINKER_FLAGS="-undefined dynamic_lookup"
Not searching for unused variables given on the command line.
-- The C compiler identification is AppleClang 8.0.0.8000042
-- The CXX compiler identification is AppleClang 8.0.0.8000042
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The ASM compiler identification is unknown
-- Found assembler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Warning: Did not find file Compiler/-ASM
-- Could NOT find LATEX (missing:  LATEX_COMPILER) 
-- Clang configured.
-- Developer mode modules added to build configuration.
-- Configuring done
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   benchlib
   benchmod
   fastcall
   ref
   sharedlib
   testlib

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /obscura/node_modules/fastcall/build
info CMD BUILD
info RUN cmake --build "/obscura/node_modules/fastcall/build" --config Release
Scanning dependencies of target dynload_s
[  2%] Building C object deps/dyncall/dynload/CMakeFiles/dynload_s.dir/dynload.c.o
[  5%] Building C object deps/dyncall/dynload/CMakeFiles/dynload_s.dir/dynload_syms.c.o
[  7%] Linking C static library libdynload_s.a
[  7%] Built target dynload_s
Scanning dependencies of target dyncall_s
[ 10%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_call.S.o
[ 12%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_vector.c.o
[ 15%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_struct.c.o
[ 17%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_api.c.o
[ 20%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_callvm.c.o
[ 23%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_callvm_base.c.o
[ 25%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_callf.c.o
[ 28%] Linking C static library libdyncall_s.a
[ 28%] Built target dyncall_s
Scanning dependencies of target dyncallback_s
[ 30%] Building C object deps/dyncall/dyncallback/CMakeFiles/dyncallback_s.dir/dyncall_callback_arch.S.o
[ 33%] Building C object deps/dyncall/dyncallback/CMakeFiles/dyncallback_s.dir/dyncall_thunk.c.o
[ 35%] Building C object deps/dyncall/dyncallback/CMakeFiles/dyncallback_s.dir/dyncall_alloc_wx.c.o
[ 38%] Building C object deps/dyncall/dyncallback/CMakeFiles/dyncallback_s.dir/dyncall_args.c.o
[ 41%] Building C object deps/dyncall/dyncallback/CMakeFiles/dyncallback_s.dir/dyncall_callback.c.o
[ 43%] Linking C static library libdyncallback_s.a
[ 43%] Built target dyncallback_s
Scanning dependencies of target ref
[ 46%] Building CXX object deps/ref-cmake/CMakeFiles/ref.dir/__/ref/src/binding.cc.o
[ 48%] Linking CXX shared library ../../Release/ref.node
[ 48%] Built target ref
Scanning dependencies of target fastcall
[ 51%] Building CXX object src/CMakeFiles/fastcall.dir/callbackimpl.cpp.o
[ 53%] Building CXX object src/CMakeFiles/fastcall.dir/dyncallbackwrapper.cpp.o
```

Seems like it could be an issue with my environment, but I'm not sure what the requirements are.

**Comment by unbornchikken** [comment]
It seems your version of clang cannot figure out that lambda argument dispatch. Which version of clang lies in Xcode 8 of yours?

On Dec 11, 2016 17:24, Arjun Mehta <notifications@github.com> wrote:

Hi there, I'm trying to install another module (NOOOCL) actually, but I'm encountering an issue with fastcall installation.

I am using MacOS Sierra with XCode 8.2.

The fastcall build fails at Building CXX object src/CMakeFiles/fastcall.dir/dyncallwrapper.cpp.o.

Scanning dependencies of target fastcall
[ 51%] Building CXX object src/CMakeFiles/fastcall.dir/callbackimpl.cpp.o
[ 53%] Building CXX object src/CMakeFiles/fastcall.dir/dyncallbackwrapper.cpp.o
[ 56%] Building CXX object src/CMakeFiles/fastcall.dir/dyncallwrapper.cpp.o
/obscura/node_modules/fastcall/src/dyncallwrapper.cpp:584:5: error: no matching function for
      call to 'CallAsync'
    CallAsync<uint64_t>(
    ^~~~~~~~~~~~~~~~~~~
/obscura/node_modules/fastcall/src/dyncallwrapper.cpp:122:13: note: candidate function
      [with T = unsigned long long] not viable: no known conversion from '(lambda at
      /obscura/node_modules/fastcall/src/dyncallwrapper.cpp:587:9)' to 'typename
      CallAsyncWorker<unsigned long long>::TConvertFunc' (aka 'v8::Local<v8::Value> (*)(unsigned long long)') for 3rd
      argument
inline void CallAsync(
            ^
1 error generated.

Full output to that point:

info TOOL Using Unix Makefiles generator.
info TOOL Using c++11 compiler standard.
info CMD CONFIGURE
info RUN cmake "/obscura/node_modules/fastcall" --no-warn-unused-cli -G"Unix Makefiles" -DCMAKE_JS_VERSION="3.3.1" -DCMAKE_BUILD_TYPE="Release" -DCMAKE_LIBRARY_OUTPUT_DIRECTORY="/obscuras/node_modules/fastcall/build/Release" -DCMAKE_JS_INC="/Users/me/.cmake-js/node-x64/v7.2.1/include/node;/obscura/node_modules/nan" -DNODE_RUNTIME="node" -DNODE_RUNTIMEVERSION="7.2.1" -DNODE_ARCH="x64" -DCMAKE_CXX_FLAGS="-std=c++11 -D_DARWIN_USE_64_BIT_INODE=1 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DBUILDING_NODE_EXTENSION -w" -DCMAKE_SHARED_LINKER_FLAGS="-undefined dynamic_lookup"
Not searching for unused variables given on the command line.
-- The C compiler identification is AppleClang 8.0.0.8000042
-- The CXX compiler identification is AppleClang 8.0.0.8000042
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The ASM compiler identification is unknown
-- Found assembler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Warning: Did not find file Compiler/-ASM
-- Could NOT find LATEX (missing:  LATEX_COMPILER)
-- Clang configured.
-- Developer mode modules added to build configuration.
-- Configuring done
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   benchlib
   benchmod
   fastcall
   ref
   sharedlib
   testlib

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /obscura/node_modules/fastcall/build
info CMD BUILD
info RUN cmake --build "/obscura/node_modules/fastcall/build" --config Release
Scanning dependencies of target dynload_s
[  2%] Building C object deps/dyncall/dynload/CMakeFiles/dynload_s.dir/dynload.c.o
[  5%] Building C object deps/dyncall/dynload/CMakeFiles/dynload_s.dir/dynload_syms.c.o
[  7%] Linking C static library libdynload_s.a
[  7%] Built target dynload_s
Scanning dependencies of target dyncall_s
[ 10%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_call.S.o
[ 12%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_vector.c.o
[ 15%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_struct.c.o
[ 17%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_api.c.o
[ 20%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_callvm.c.o
[ 23%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_callvm_base.c.o
[ 25%] Building C object deps/dyncall/dyncall/CMakeFiles/dyncall_s.dir/dyncall_callf.c.o
[ 28%] Linking C static library libdyncall_s.a
[ 28%] Built target dyncall_s
Scanning dependencies of target dyncallback_s
[ 30%] Building C object deps/dyncall/dyncallback/CMakeFiles/dyncallback_s.dir/dyncall_callback_arch.S.o
[ 33%] Building C object deps/dyncall/dyncallback/CMakeFiles/dyncallback_s.dir/dyncall_thunk.c.o
[ 35%] Building C object deps/dyncall/dyncallback/CMakeFiles/dyncallback_s.dir/dyncall_alloc_wx.c.o
[ 38%] Building C object deps/dyncall/dyncallback/CMakeFiles/dyncallback_s.dir/dyncall_args.c.o
[ 41%] Building C object deps/dyncall/dyncallback/CMakeFiles/dyncallback_s.dir/dyncall_callback.c.o
[ 43%] Linking C static library libdyncallback_s.a
[ 43%] Built target dyncallback_s
Scanning dependencies of target ref
[ 46%] Building CXX object deps/ref-cmake/CMakeFiles/ref.dir/__/ref/src/binding.cc.o
[ 48%] Linking CXX shared library ../../Release/ref.node
[ 48%] Built target ref
Scanning dependencies of target fastcall
[ 51%] Building CXX object src/CMakeFiles/fastcall.dir/callbackimpl.cpp.o
[ 53%] Building CXX object src/CMakeFiles/fastcall.dir/dyncallbackwrapper.cpp.o

Seems like it could be an issue with my environment, but I'm not sure what the requirements are.

**Comment by arjunmehta** [comment]
@unbornchikken Here's the output from `clang -v`.

```bash
clang -v
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin16.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Seems to match what cmake is detecting
`The CXX compiler identification is AppleClang 8.0.0.8000042`

**Comment by arjunmehta** [comment]
I just tried it again on a clone.

Tried with `Node 4.7.0` & `npm 2.15.11` & `node-gyp 3.4.0`
Seems like it is not giving the same Async error. But cmake is outputting the following:

```bash
-- The C compiler identification is AppleClang 8.0.0.8000042
-- The CXX compiler identification is AppleClang 8.0.0.8000042
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The ASM compiler identification is unknown
-- Found assembler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Warning: Did not find file Compiler/-ASM
-- Could NOT find LATEX (missing:  LATEX_COMPILER) 
-- Clang configured.
You have called ADD_LIBRARY for library ref without any source files. This typically indicates a problem with your CMakeLists.txt file
-- Developer mode modules added to build configuration.
-- Configuring done
CMake Error: CMake can not determine linker language for target: ref
CMake Error: Cannot determine link language for target "ref".
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   benchlib
   benchmod
   fastcall
   sharedlib
   testlib

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /Users/arjun/Working/testing/fastcall/build
```

**Comment by unbornchikken** [comment]
I don't own a Mac. But it definitely builds and works on Linux + clang latest. I've a plan for Travis CI integration, that will invoke build and test for each release on OSX too, and issues like this are gonna be eliminated. Let me finish https://github.com/cmake-js/fastcall/issues/7 first.

**Comment by unbornchikken** [comment]
And I have plan for implementing this: https://github.com/cmake-js/cmake-js/issues/70. That will be a huge win for issues like this, because binary distribution means there won't be any compiling during npm install.

**Comment by unbornchikken** [comment]
Fixed as of 0.1.29.

Please note that this repo contains git submodules. So, before trying to build the clone, you gotta do:

```bash
git submodule update --init --recursive
```

I'm gonna write a contributor FAQ with more on this once I have some spare time.

**Comment by arjunmehta** [comment]
@unbornchikken Looks like it is installing okay with the latest patch. Thanks!
